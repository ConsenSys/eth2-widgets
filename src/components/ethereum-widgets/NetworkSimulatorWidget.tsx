import React, { useState, useEffect } from 'react'
import { Collapse, Icon, Button } from 'antd';
import { CartesianGrid, XAxis, YAxis, Label, AreaChart, Area, Legend, Tooltip } from 'recharts';

import { MINIMAL_ETH_STAKED, MINIMAL_AVERAGE_PERCENTAGE_VALIDATORS_ONLINE, BIGGEST_PROBABILITY, NET_REWARDS_COLOUR, REWARDS_COLOUR, PENALTIES_COLOUR } from './constants'
import { EthStakedSlider, OnlineProbabilitySlider, HonestProbabilitySlider } from './Sliders'
import { buildFakeSimulationData } from './faker';
import { ControlsContainer, ChartContainer } from './Containers';

import "./styles.css"

const { Panel } = Collapse;

export const NetworkSimulatorWidget: React.FC = () => {
    const [isReady, setIsReady] = useState(false)
    const [wasmClient, setWasmClient] = useState(undefined)
    const [ethStaked, setEthStaked] = useState(MINIMAL_ETH_STAKED)
    const [onlineProbability, setOnlineProbability] = useState(BIGGEST_PROBABILITY)
    const [honestProbability, setHonestProbability] = useState(BIGGEST_PROBABILITY)
    const [averagePercentageOfValidatorOnline, setAveragePercentageOfValidatorOnline] = useState(MINIMAL_AVERAGE_PERCENTAGE_VALIDATORS_ONLINE)
    const [validatorAnnualInterest, setValidatorAnnualInterest] = useState(0)
    const [showAdvanceOptions, setShowAdvanceOptions] = useState(false)
    const [netRewardsActive, setNetRewardsActive] = useState(true)
    const [rewardsActive, setRewardsActive] = useState(true)
    const [penaltiesActive, setPenaltiesActive] = useState(true)
    const [isRunning, setIsRunning] = useState(false)

    useEffect(() => {
        const loadClient = async () => {
            import("eth2-simulator")
                .then(module => {
                    console.log("MODULE", module.get_validator_reward)
                    setWasmClient(module)
                    setIsReady(true)
                    console.log("WASM MODULE LOADED")
                }).catch((error) => {
                    console.log("Error loading the module")
                });
        }
        loadClient()
            .catch(e => {
                console.log("ERROR")
            })
    }, [])

    useEffect(() => {
        if (isReady) {
            console.log("WASM CLIENT", wasmClient)
            console.log("USE EFFECT RUNNING", ethStaked, averagePercentageOfValidatorOnline)
            const result = wasmClient.get_validator_reward(ethStaked, averagePercentageOfValidatorOnline);
            setValidatorAnnualInterest(result)
        }
    });

    const handleLegendClick = ({ value }) => {
        if (value === "Rewards") {
            setRewardsActive(!rewardsActive)
        }
        else if (value === "Net rewards") {
            setNetRewardsActive(!netRewardsActive)
        }
        else if (value === "Penalties") {
            setPenaltiesActive(!penaltiesActive)
        }
    }

    const runSimulation = () => {
        setIsRunning(true)
    }

    return (
        <div className="widget-container">
            <ControlsContainer title="Network Simulation">
                <EthStakedSlider
                    onChange={(e) => {
                        console.log("New Eth Staked Value", e)
                        // call rust module
                        setEthStaked(e)
                    }}
                    initialValue={MINIMAL_ETH_STAKED}
                />
                <Collapse
                    accordion={true}
                    expandIcon={() => <Icon type="setting" />}
                    onChange={() => setShowAdvanceOptions(!showAdvanceOptions)}>
                    <Panel header="Advance options" key="advanced">
                        <OnlineProbabilitySlider
                            initialValue={BIGGEST_PROBABILITY}
                            onChange={(e) => {
                                console.log("New online probability value", e)
                                // call rust module
                                setOnlineProbability(e)
                            }}
                        />
                        <HonestProbabilitySlider
                            initialValue={BIGGEST_PROBABILITY}
                            onChange={(e) => {
                                console.log("New honest probability value", e)
                                // call rust module
                                setHonestProbability(e)
                            }}
                        />
                    </Panel>
                </Collapse>
                <Button onClick={runSimulation} type="primary" block={true} loading={isRunning}>{isRunning ? "" : "Run"}</Button>
            </ControlsContainer>

            <ChartContainer isLoading={isRunning}>
                <AreaChart width={800} height={400} data={buildFakeSimulationData()}>
                    <Legend onClick={handleLegendClick} verticalAlign="top" align="right" height={36} />
                    <Tooltip formatter={(value, name, props) => (`${value}%`)} />
                    <defs>
                        <linearGradient id="colorUv" x1="0" y1="0" x2="0" y2="1">
                            <stop offset="5%" stopColor={NET_REWARDS_COLOUR} stopOpacity={0.8} />
                            <stop offset="95%" stopColor={NET_REWARDS_COLOUR} stopOpacity={0} />
                        </linearGradient>
                        <linearGradient id="colorPv" x1="0" y1="0" x2="0" y2="1">
                            <stop offset="5%" stopColor={REWARDS_COLOUR} stopOpacity={0.8} />
                            <stop offset="95%" stopColor={REWARDS_COLOUR} stopOpacity={0} />
                        </linearGradient>
                        <linearGradient id="colorQv" x1="0" y1="0" x2="0" y2="1">
                            <stop offset="5%" stopColor={PENALTIES_COLOUR} stopOpacity={0.8} />
                            <stop offset="95%" stopColor={PENALTIES_COLOUR} stopOpacity={0} />
                        </linearGradient>
                    </defs>

                    <Area type="monotone" dataKey={rewardsActive ? "rewards" : ""} stroke={REWARDS_COLOUR} fillOpacity={1} fill="url(#colorPv)" name="Rewards" />
                    <Area type="monotone" dataKey={netRewardsActive ? "net_rewards" : ""} stroke={NET_REWARDS_COLOUR} fillOpacity={1} fill="url(#colorUv)" name="Net rewards" />
                    <Area type="monotone" dataKey={penaltiesActive ? "penalties" : ""} stroke={PENALTIES_COLOUR} fillOpacity={1} fill="url(#colorQv)" name="Penalties" />

                    <CartesianGrid stroke="#D5DCE4" strokeDasharray="5 5" />

                    <XAxis stroke="#97A4BA" dataKey="time">
                        <Label style={{ textAnchor: 'middle', fontSize: '80%' }} value="Months" offset={0} fill="#97A4BA" position="insideBottom" />
                    </XAxis>

                    <YAxis stroke="#97A4BA" datakey="net_rewards" unit="%" name="net_rewards" viewBox={200} />

                </AreaChart>
            </ChartContainer>
            {/* <div className="chart-container">
                {isRunning ? <div style={{ width: "100%", height: "100%", display: "flex", alignItems: "center", justifyContent: "center", border: "1px solid #FAFAFC" }}><Spin tip="Running simulation..." /></div>
                    :
                    <LineChart width={800} height={400} data={buildFakeSimulationData()}>
                        <Legend onClick={handleLegendClick} verticalAlign="top" align="right" height={36} />
                        <Line type="monotone" strokeWidth={rewardsActive ? 1 : 0} dataKey="rewards" stroke={REWARDS_COLOUR} name="Rewards" />

                        <Line type="monotone" strokeWidth={netRewardsActive ? 1 : 0} dataKey="net_rewards" stroke={NET_REWARDS_COLOUR} name="Net rewards" />

                        <Line type="monotone" strokeWidth={penaltiesActive ? 1 : 0} dataKey="penalties" stroke={PENALTIES_COLOUR} name="Penalties" />
                        <CartesianGrid stroke="#D5DCE4" strokeDasharray="5 5" />
                        <XAxis stroke="#97A4BA" dataKey="time">
                            <Label style={{ textAnchor: 'middle', fontSize: '80%' }} value="Months" offset={0} fill="#97A4BA" position="insideBottom" />
                        </XAxis>
                        <YAxis stroke="#97A4BA" datakey="net_rewards" unit="%" name="net_rewards" viewBox={200} />
                    </LineChart>}
            </div> */}
        </div>
    )
}