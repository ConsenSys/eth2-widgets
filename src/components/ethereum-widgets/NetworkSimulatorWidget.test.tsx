import React from 'react';
import { render, fireEvent, waitForElement, getByText } from '@testing-library/react';
import { NetworkSimulatorWidget } from './NetworkSimulatorWidget'
import '@testing-library/jest-dom/extend-expect';

describe('Network simulator Widget Component', () => {

    const renderWidget = () => {
        return render(<NetworkSimulatorWidget />)
    }

    test('that the ethereum widget renders correctly', () => {
        const expectedWidgetTitle = 'Network Simulation'
        const { getByText, getByTestId, debug } = renderWidget()

        expect(getByText(expectedWidgetTitle)).toBeDefined()
        expect(getByTestId('widget-title')).toBeDefined()
        expect(getByTestId('widget-title')).toBeInTheDocument()
    })

    test('that the ethereum widget snapshot is correct', () => {
        const expectedWidgetTitle = 'Network Simulation'
        const { container, debug } = renderWidget()

        expect(container).toMatchSnapshot()
    })
})