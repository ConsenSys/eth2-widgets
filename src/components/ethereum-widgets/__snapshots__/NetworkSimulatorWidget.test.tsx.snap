// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Network simulator Widget Component that the ethereum widget snapshot is correct 1`] = `
<div>
  <div
    class="widget-container"
  >
    <div
      class="controls-container"
    >
      <h3
        data-testid="widget-title"
      >
        Network Simulation
      </h3>
      <div
        style="margin-top: 1em; width: 100%;"
      >
        <h4>
          Total Eth Staked
        </h4>
        <div
          class="ant-slider ant-slider-with-marks"
          style="padding-bottom: 1em;"
        >
          <div
            class="ant-slider-rail"
          />
          <div
            class="ant-slider-track"
            style="left: 0%; width: 0%;"
          />
          <div
            class="ant-slider-step"
          >
            <span
              class="ant-slider-dot ant-slider-dot-active"
              style="left: 0%;"
            />
            <span
              class="ant-slider-dot"
              style="left: 100%;"
            />
          </div>
          <div
            aria-disabled="false"
            aria-valuemax="10"
            aria-valuemin="1"
            aria-valuenow="1"
            class="ant-slider-handle"
            role="slider"
            style="left: 0%; transform: translateX(-50%);"
            tabindex="0"
          />
          <div
            class="ant-slider-mark"
          >
            <span
              class="ant-slider-mark-text ant-slider-mark-text-active"
              style="transform: translateX(-50%); left: 0%; color: rgb(103, 117, 141);"
            >
              <p
                class="legend-slider"
              >
                1 Million
              </p>
            </span>
            <span
              class="ant-slider-mark-text"
              style="transform: translateX(-50%); left: 100%; color: rgb(103, 117, 141);"
            >
              <p
                class="legend-slider"
              >
                10 Millions
              </p>
            </span>
          </div>
        </div>
      </div>
      <div
        class="ant-collapse ant-collapse-icon-position-left"
        role="tablist"
      >
        <div
          class="ant-collapse-item"
        >
          <div
            aria-expanded="false"
            class="ant-collapse-header"
            role="tab"
            tabindex="0"
          >
            <i
              aria-label="icon: setting"
              class="anticon anticon-setting ant-collapse-arrow"
            >
              <svg
                aria-hidden="true"
                class=""
                data-icon="setting"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M924.8 625.7l-65.5-56c3.1-19 4.7-38.4 4.7-57.8s-1.6-38.8-4.7-57.8l65.5-56a32.03 32.03 0 0 0 9.3-35.2l-.9-2.6a443.74 443.74 0 0 0-79.7-137.9l-1.8-2.1a32.12 32.12 0 0 0-35.1-9.5l-81.3 28.9c-30-24.6-63.5-44-99.7-57.6l-15.7-85a32.05 32.05 0 0 0-25.8-25.7l-2.7-.5c-52.1-9.4-106.9-9.4-159 0l-2.7.5a32.05 32.05 0 0 0-25.8 25.7l-15.8 85.4a351.86 351.86 0 0 0-99 57.4l-81.9-29.1a32 32 0 0 0-35.1 9.5l-1.8 2.1a446.02 446.02 0 0 0-79.7 137.9l-.9 2.6c-4.5 12.5-.8 26.5 9.3 35.2l66.3 56.6c-3.1 18.8-4.6 38-4.6 57.1 0 19.2 1.5 38.4 4.6 57.1L99 625.5a32.03 32.03 0 0 0-9.3 35.2l.9 2.6c18.1 50.4 44.9 96.9 79.7 137.9l1.8 2.1a32.12 32.12 0 0 0 35.1 9.5l81.9-29.1c29.8 24.5 63.1 43.9 99 57.4l15.8 85.4a32.05 32.05 0 0 0 25.8 25.7l2.7.5a449.4 449.4 0 0 0 159 0l2.7-.5a32.05 32.05 0 0 0 25.8-25.7l15.7-85a350 350 0 0 0 99.7-57.6l81.3 28.9a32 32 0 0 0 35.1-9.5l1.8-2.1c34.8-41.1 61.6-87.5 79.7-137.9l.9-2.6c4.5-12.3.8-26.3-9.3-35zM788.3 465.9c2.5 15.1 3.8 30.6 3.8 46.1s-1.3 31-3.8 46.1l-6.6 40.1 74.7 63.9a370.03 370.03 0 0 1-42.6 73.6L721 702.8l-31.4 25.8c-23.9 19.6-50.5 35-79.3 45.8l-38.1 14.3-17.9 97a377.5 377.5 0 0 1-85 0l-17.9-97.2-37.8-14.5c-28.5-10.8-55-26.2-78.7-45.7l-31.4-25.9-93.4 33.2c-17-22.9-31.2-47.6-42.6-73.6l75.5-64.5-6.5-40c-2.4-14.9-3.7-30.3-3.7-45.5 0-15.3 1.2-30.6 3.7-45.5l6.5-40-75.5-64.5c11.3-26.1 25.6-50.7 42.6-73.6l93.4 33.2 31.4-25.9c23.7-19.5 50.2-34.9 78.7-45.7l37.9-14.3 17.9-97.2c28.1-3.2 56.8-3.2 85 0l17.9 97 38.1 14.3c28.7 10.8 55.4 26.2 79.3 45.8l31.4 25.8 92.8-32.9c17 22.9 31.2 47.6 42.6 73.6L781.8 426l6.5 39.9zM512 326c-97.2 0-176 78.8-176 176s78.8 176 176 176 176-78.8 176-176-78.8-176-176-176zm79.2 255.2A111.6 111.6 0 0 1 512 614c-29.9 0-58-11.7-79.2-32.8A111.6 111.6 0 0 1 400 502c0-29.9 11.7-58 32.8-79.2C454 401.6 482.1 390 512 390c29.9 0 58 11.6 79.2 32.8A111.6 111.6 0 0 1 624 502c0 29.9-11.7 58-32.8 79.2z"
                />
              </svg>
            </i>
            Advance options
          </div>
        </div>
      </div>
      <button
        class="ant-btn ant-btn-primary ant-btn-block"
        type="button"
      >
        <span>
          Run
        </span>
      </button>
    </div>
    <div
      class="chart-container"
    >
      <div
        class="recharts-wrapper"
        style="position: relative; cursor: default; width: 800px; height: 400px;"
      >
        <svg
          class="recharts-surface"
          height="400"
          version="1.1"
          viewBox="0 0 800 400"
          width="800"
        >
          <defs>
            <clippath
              id="recharts11-clip"
            >
              <rect
                height="360"
                width="730"
                x="65"
                y="5"
              />
            </clippath>
          </defs>
          <defs>
            <lineargradient
              id="colorUv"
              x1="0"
              x2="0"
              y1="0"
              y2="1"
            >
              <stop
                offset="5%"
                stop-color="#3794FC"
                stop-opacity="0.8"
              />
              <stop
                offset="95%"
                stop-color="#3794FC"
                stop-opacity="0"
              />
            </lineargradient>
            <lineargradient
              id="colorPv"
              x1="0"
              x2="0"
              y1="0"
              y2="1"
            >
              <stop
                offset="5%"
                stop-color="#9D60FB"
                stop-opacity="0.8"
              />
              <stop
                offset="95%"
                stop-color="#9D60FB"
                stop-opacity="0"
              />
            </lineargradient>
            <lineargradient
              id="colorQv"
              x1="0"
              x2="0"
              y1="0"
              y2="1"
            >
              <stop
                offset="5%"
                stop-color="#FF0000"
                stop-opacity="0.8"
              />
              <stop
                offset="95%"
                stop-color="#FF0000"
                stop-opacity="0"
              />
            </lineargradient>
          </defs>
          <g
            class="recharts-cartesian-grid"
          >
            <g
              class="recharts-cartesian-grid-horizontal"
            >
              <line
                fill="none"
                height="360"
                offset="[object Object]"
                stroke="#D5DCE4"
                stroke-dasharray="5 5"
                width="730"
                x="65"
                x1="65"
                x2="795"
                y="5"
                y1="365"
                y2="365"
              />
              <line
                fill="none"
                height="360"
                offset="[object Object]"
                stroke="#D5DCE4"
                stroke-dasharray="5 5"
                width="730"
                x="65"
                x1="65"
                x2="795"
                y="5"
                y1="275"
                y2="275"
              />
              <line
                fill="none"
                height="360"
                offset="[object Object]"
                stroke="#D5DCE4"
                stroke-dasharray="5 5"
                width="730"
                x="65"
                x1="65"
                x2="795"
                y="5"
                y1="185"
                y2="185"
              />
              <line
                fill="none"
                height="360"
                offset="[object Object]"
                stroke="#D5DCE4"
                stroke-dasharray="5 5"
                width="730"
                x="65"
                x1="65"
                x2="795"
                y="5"
                y1="95"
                y2="95"
              />
              <line
                fill="none"
                height="360"
                offset="[object Object]"
                stroke="#D5DCE4"
                stroke-dasharray="5 5"
                width="730"
                x="65"
                x1="65"
                x2="795"
                y="5"
                y1="5"
                y2="5"
              />
            </g>
            <g
              class="recharts-cartesian-grid-vertical"
            >
              <line
                fill="none"
                height="360"
                offset="[object Object]"
                stroke="#D5DCE4"
                stroke-dasharray="5 5"
                width="730"
                x="65"
                x1="65"
                x2="65"
                y="5"
                y1="5"
                y2="365"
              />
              <line
                fill="none"
                height="360"
                offset="[object Object]"
                stroke="#D5DCE4"
                stroke-dasharray="5 5"
                width="730"
                x="65"
                x1="795"
                x2="795"
                y="5"
                y1="5"
                y2="365"
              />
            </g>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
          >
            <line
              class="recharts-yAxis yAxis"
              fill="none"
              height="360"
              name="net_rewards"
              orientation="left"
              stroke="#97A4BA"
              type="number"
              viewBox="[object Object]"
              width="60"
              x="5"
              x1="65"
              x2="65"
              y="5"
              y1="5"
              y2="365"
            />
            <g
              class="recharts-cartesian-axis-ticks"
            >
              <g
                class="recharts-layer recharts-cartesian-axis-tick"
              >
                <line
                  class="recharts-yAxis yAxis"
                  fill="none"
                  height="360"
                  name="net_rewards"
                  orientation="left"
                  stroke="#97A4BA"
                  type="number"
                  viewBox="[object Object]"
                  width="60"
                  x="5"
                  x1="59"
                  x2="65"
                  y="5"
                  y1="365"
                  y2="365"
                />
                <text
                  class="recharts-text recharts-cartesian-axis-tick-value"
                  fill="#97A4BA"
                  height="360"
                  name="net_rewards"
                  orientation="left"
                  stroke="none"
                  text-anchor="end"
                  type="number"
                  viewBox="[object Object]"
                  width="60"
                  x="57"
                  y="365"
                >
                  <tspan
                    dy="0.355em"
                    x="57"
                  >
                    0%
                  </tspan>
                </text>
              </g>
              <g
                class="recharts-layer recharts-cartesian-axis-tick"
              >
                <line
                  class="recharts-yAxis yAxis"
                  fill="none"
                  height="360"
                  name="net_rewards"
                  orientation="left"
                  stroke="#97A4BA"
                  type="number"
                  viewBox="[object Object]"
                  width="60"
                  x="5"
                  x1="59"
                  x2="65"
                  y="5"
                  y1="275"
                  y2="275"
                />
                <text
                  class="recharts-text recharts-cartesian-axis-tick-value"
                  fill="#97A4BA"
                  height="360"
                  name="net_rewards"
                  orientation="left"
                  stroke="none"
                  text-anchor="end"
                  type="number"
                  viewBox="[object Object]"
                  width="60"
                  x="57"
                  y="275"
                >
                  <tspan
                    dy="0.355em"
                    x="57"
                  >
                    5%
                  </tspan>
                </text>
              </g>
              <g
                class="recharts-layer recharts-cartesian-axis-tick"
              >
                <line
                  class="recharts-yAxis yAxis"
                  fill="none"
                  height="360"
                  name="net_rewards"
                  orientation="left"
                  stroke="#97A4BA"
                  type="number"
                  viewBox="[object Object]"
                  width="60"
                  x="5"
                  x1="59"
                  x2="65"
                  y="5"
                  y1="185"
                  y2="185"
                />
                <text
                  class="recharts-text recharts-cartesian-axis-tick-value"
                  fill="#97A4BA"
                  height="360"
                  name="net_rewards"
                  orientation="left"
                  stroke="none"
                  text-anchor="end"
                  type="number"
                  viewBox="[object Object]"
                  width="60"
                  x="57"
                  y="185"
                >
                  <tspan
                    dy="0.355em"
                    x="57"
                  >
                    10%
                  </tspan>
                </text>
              </g>
              <g
                class="recharts-layer recharts-cartesian-axis-tick"
              >
                <line
                  class="recharts-yAxis yAxis"
                  fill="none"
                  height="360"
                  name="net_rewards"
                  orientation="left"
                  stroke="#97A4BA"
                  type="number"
                  viewBox="[object Object]"
                  width="60"
                  x="5"
                  x1="59"
                  x2="65"
                  y="5"
                  y1="95"
                  y2="95"
                />
                <text
                  class="recharts-text recharts-cartesian-axis-tick-value"
                  fill="#97A4BA"
                  height="360"
                  name="net_rewards"
                  orientation="left"
                  stroke="none"
                  text-anchor="end"
                  type="number"
                  viewBox="[object Object]"
                  width="60"
                  x="57"
                  y="95"
                >
                  <tspan
                    dy="0.355em"
                    x="57"
                  >
                    15%
                  </tspan>
                </text>
              </g>
              <g
                class="recharts-layer recharts-cartesian-axis-tick"
              >
                <line
                  class="recharts-yAxis yAxis"
                  fill="none"
                  height="360"
                  name="net_rewards"
                  orientation="left"
                  stroke="#97A4BA"
                  type="number"
                  viewBox="[object Object]"
                  width="60"
                  x="5"
                  x1="59"
                  x2="65"
                  y="5"
                  y1="5"
                  y2="5"
                />
                <text
                  class="recharts-text recharts-cartesian-axis-tick-value"
                  fill="#97A4BA"
                  height="360"
                  name="net_rewards"
                  orientation="left"
                  stroke="none"
                  text-anchor="end"
                  type="number"
                  viewBox="[object Object]"
                  width="60"
                  x="57"
                  y="5"
                >
                  <tspan
                    dy="0.355em"
                    x="57"
                  >
                    20%
                  </tspan>
                </text>
              </g>
            </g>
          </g>
        </svg>
        <div
          class="recharts-legend-wrapper"
          style="position: absolute; width: 790px; height: 36px; right: 5px; top: 5px;"
        >
          <ul
            class="recharts-default-legend"
            style="padding: 0px; margin: 0px; text-align: right;"
          >
            <li
              class="recharts-legend-item legend-item-0"
              style="display: inline-block; margin-right: 10px;"
            >
              <svg
                class="recharts-surface"
                height="14"
                style="display: inline-block; vertical-align: middle; margin-right: 4px;"
                version="1.1"
                viewBox="0 0 32 32"
                width="14"
              >
                <path
                  class="recharts-legend-icon"
                  d="M0,16h10.666666666666666
            A5.333333333333333,5.333333333333333,0,1,1,21.333333333333332,16
            H32M21.333333333333332,16
            A5.333333333333333,5.333333333333333,0,1,1,10.666666666666666,16"
                  fill="none"
                  stroke="#9D60FB"
                  stroke-width="4"
                />
              </svg>
              <span
                class="recharts-legend-item-text"
              >
                Rewards
              </span>
            </li>
            <li
              class="recharts-legend-item legend-item-1"
              style="display: inline-block; margin-right: 10px;"
            >
              <svg
                class="recharts-surface"
                height="14"
                style="display: inline-block; vertical-align: middle; margin-right: 4px;"
                version="1.1"
                viewBox="0 0 32 32"
                width="14"
              >
                <path
                  class="recharts-legend-icon"
                  d="M0,16h10.666666666666666
            A5.333333333333333,5.333333333333333,0,1,1,21.333333333333332,16
            H32M21.333333333333332,16
            A5.333333333333333,5.333333333333333,0,1,1,10.666666666666666,16"
                  fill="none"
                  stroke="#3794FC"
                  stroke-width="4"
                />
              </svg>
              <span
                class="recharts-legend-item-text"
              >
                Net rewards
              </span>
            </li>
            <li
              class="recharts-legend-item legend-item-2"
              style="display: inline-block; margin-right: 10px;"
            >
              <svg
                class="recharts-surface"
                height="14"
                style="display: inline-block; vertical-align: middle; margin-right: 4px;"
                version="1.1"
                viewBox="0 0 32 32"
                width="14"
              >
                <path
                  class="recharts-legend-icon"
                  d="M0,16h10.666666666666666
            A5.333333333333333,5.333333333333333,0,1,1,21.333333333333332,16
            H32M21.333333333333332,16
            A5.333333333333333,5.333333333333333,0,1,1,10.666666666666666,16"
                  fill="none"
                  stroke="#FF0000"
                  stroke-width="4"
                />
              </svg>
              <span
                class="recharts-legend-item-text"
              >
                Penalties
              </span>
            </li>
          </ul>
        </div>
        <div
          class="recharts-tooltip-wrapper"
          style="transform: translate(undefinedpx, undefinedpx); pointer-events: none; visibility: hidden; position: absolute; top: 0px;"
        >
          <div
            class="recharts-default-tooltip"
            style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); border: 1px solid #ccc; white-space: nowrap;"
          >
            <p
              class="recharts-tooltip-label"
              style="margin: 0px;"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
